python/mlc_chat/support/tensor_parallel.py

```diff
@@ -8,208 +7,91 @@
 
 
 @dataclasses.dataclass
-class Row:
-    """Shard a 2D tensor by its rows."""
+class ShardSingleDim:
+    """Shard a tensor by one of its dimension."""
 
     name: str
-    row: int
-    col: int
+    dim: int
 
     def gen_tir(self, shards: int, weight: nn.Tensor) -> tir.PrimFunc:
         """Generate a TIR function that shards the weight tensor by its rows."""
-        assert weight.shape == [self.row, self.col]
-        w = te.placeholder([self.row * shards, self.col], weight.dtype, name="w")
-        o = topi.reshape(w, (shards, self.row, self.col))
+        orig_shape = [
+            weight.shape[i] if i != self.dim else weight.shape[self.dim] * shards
+            for i in range(len(weight.shape))
+        ]
+        reshape_shape = [
+            *weight.shape[: self.dim],
+            shards,
+            weight.shape[self.dim],
+            *weight.shape[self.dim + 1 :],
+        ]
+        transpose_index = [self.dim, *range(self.dim), *range(self.dim + 1, len(weight.shape) + 1)]
+        w = te.placeholder(orig_shape, weight.dtype, name="w")
+        reshape = topi.reshape(w, reshape_shape)
+        o = topi.transpose(reshape, transpose_index)
         func = te.create_prim_func([w, o])
         return func
 
     def gen_shard_info(self, shards: int, weight: nn.Tensor) -> Dict[str, Any]:
         """Generate shard info for this sharding strategy."""
-        assert weight.shape == [self.row, self.col]
         return {
             "func_name": self.name,
-            "out_shape": (shards, self.row, self.col),
+            "out_shape": (shards, *weight.shape),
             "out_dtype": weight.dtype,
         }
 
 
 @dataclasses.dataclass
-class Col:
-    """Shard a 2D tensor by its columns."""
+class ShardSingleDimSeg:
```

Comment:
Is `ShardSingleDim` a special case of `ShardSingleDimSeg`? Perhaps it's something we could further unify

Reply:
Yes `ShardSingleDim` is a special case of `ShardSingleDimSeg` when `segs=[shape[dim]]`. However I'm concerned that if we unify them, it would be harder to use because ShardSingleDimSeg requires `segs` as input, while ShardSingleDim doesn't require it. Also ShardSingleDim is more common and more understandable, so I think we should keep both `ShardSingleDim` and `ShardSingleDimSeg`

Reply:
Instead, do you think we could provide a unified `ShardSingleDim` whose constructor signature looks like:

```python
def __init__(
  self,
  shape: List[int],
  dim: int,
  segs: Optional[List[int]] = None,
): ...
```

and if `segs` is None, it defaults to `shape[dim]`.

Reply:
This seems better to me. One minor comment, the signature does not need a `shape` parameter. 
```
def __init__(
  self,
  dim: int,
  segs: Optional[List[int]] = None,
): ...
``` 
should be good

Reply:
got it. sounds good!

Reply:
The unified `ShardSingleDim` looks great!